File: src/2348_number_of_zero_filled_subarrays.js
/**
 * 2348. Number of Zero-Filled Subarrays
 * -----------------------
 * link: https://leetcode.com/problems/number-of-zero-filled-subarrays/
 *
 * description:
 * length => no of sub arrays
 * '0' => 1
 * '00' => 3
 * '000' => 6
 * '0000' => 10
 * for each zero we found count++ and adding count to result
 * if the element not zero => count = 0
 *
 * time : O(n)
 * space : O(1)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */

let zeroFilledSubarray = function (nums) {
    let res = nums.filter((num) => num === 0).length;
    if (res === 0) {
        return 0;
    }

    let r = 0;
    const l = nums.length;
    while (r < l) {
        const tempSubarray = [];
        while (r < l && nums[r] === 0) {
            tempSubarray.push(nums[r]);
            r += 1;
        }
        if (tempSubarray.length > 1) {
            const tempCount = (tempSubarray.length * (tempSubarray.length - 1)) / 2;
            res += Math.trunc(tempCount);
        }

        r += 1;
    }
    return res;
};

module.exports = { zeroFilledSubarray };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(zeroFilledSubarray([1, 3, 0, 0, 2, 0, 0, 4]), 6);
    assert.deepStrictEqual(zeroFilledSubarray([0, 0, 0, 2, 0, 0]), 9);
}

File: src/2348_number_of_zero_filled_subarrays.rs
struct Solution;

impl Solution {
    pub fn zero_filled_subarray(nums: Vec<i32>) -> i32 {
        let mut res = nums.iter().filter(|&&value| value == 0).count() as i32;
        if res == 0 {
            return 0;
        }

        let mut r = 0;
        let l = nums.len();
        while r < l {
            let mut temp_subarray = Vec::new();
            while r < l && nums[r] == 0 {
                temp_subarray.push(nums[r]);
                r += 1;
            }
            if temp_subarray.len() > 1 {
                let temp_count = temp_subarray.len() * (temp_subarray.len() - 1) / 2;
                res += temp_count as i32;
            }
            r += 1;
        }

        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::zero_filled_subarray(vec![1, 3, 0, 0, 2, 0, 0, 4]),
            6
        );
        assert_eq!(Solution::zero_filled_subarray(vec![0, 0, 0, 2, 0, 0]), 9);
    }
}
